### PR TITLE
Don't marshal empty byte or uint8 slice as null

### PR DESCRIPTION
**Pull request by nikhita**
Fixes https://github.com/json-iterator/go/issues/272

`[]byte` or `[]uint8` are encoded as base-64 encoded strings. So, only nil slices of these types should get marshalled as `null`, and non-nil empty slices should get marshalled as `""`.

This restores compatibility with the standard library.

I verified that this patch works locally. Even though tests for these exist in https://github.com/json-iterator/go/blob/master/type_tests/slice_test.go, I couldn't reproduce a case where the test case was fuzzed to an empty `[]byte` or `[]uint8` slice... :confused:

/cc @taowen @thockin 
fyi @liggitt @dims @neolit123

### CODE REVIEW

**Comment by nikhita** [comment]
cc @sttts

**Comment by dims** [comment]
@nikhita i was playing with this snippet - https://play.golang.org/p/IOXp4klmNvU

before applying your patch, both "case 1" and "case 2" end up printing `marshal : "null"`
after applying your patch, "case 1" prints `marshal : "null"` and "case 2" prints `marshal : "\"\""`

does this help?

File: reflect_native.go [hunk 1]
```diff
@@ -432,11 +432,11 @@ func (codec *base64Codec) Decode(ptr unsafe.Pointer, iter *Iterator) {
...
-	src := *((*[]byte)(ptr))
-	if len(src) == 0 {
+	if codec.sliceType.UnsafeIsNil(ptr) {
 		stream.WriteNil()
 		return
 	}
+	src := *((*[]byte)(ptr))
 	encoding := base64.StdEncoding
 	stream.writeByte('"')
 	size := encoding.EncodedLen(len(src))
```
Comment (liggitt) [inline reflect_native.go]: you can skip lines 442-445 `if len(src) == 0`

**Comment by nikhita** [comment]
Updated with tests. Turned out that there are special tests for encoding byte arrays.

**Comment by dims** [comment]
@nikhita do we have to add the test for the `var mySlice1 []byte         // Case 1`? where it's marshalled to `null`?

**Comment by nikhita** [comment]
@dims Done.

**Comment by thockin** [comment]
LGTM.  I try to be judicious in my authority on this repo, but this seems like an important edge case.
